lib/components/searchbox.js

```diff
@@ -0,0 +1,190 @@
+
+import Component from '../component';
+
+class Searchform extends Component {
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      searchTerm: '',
+      searchError: false
+    };
+
+    this.handleSubmit = this.handleSubmit.bind(this);
+    this.handleSearchTermChange = this.handleSearchTermChange.bind(this);
+  }
+
+  handleSubmit(e) {
+    e.preventDefault();
+
+    // If a submit button was selected, return focus back to the input box
+    this.searchBox.focus();
+
+    const term = this.props.terms && this.props.terms.getActiveTerm();
+    if (term) {
+      const searchTerm = this.state.searchTerm;
+      const options = {
+        dir: this.state.dir
+      };
+      const searchResult = term.find(searchTerm, options);
+      this.setState({searchError: !searchResult});
+    } else {
+      console.error('No active terminal found to search in!');
+    }
+  }
+
+  handleSearchTermChange(e) {
+    this.setState({
+      searchTerm: e.target.value || '',
+      searchError: false
+    });
+  }
+
+  componentDidMount() {
+    this.keys = new Mousetrap(this.form);
+
+    // Search upwards while the shift key is down
+    this.keys.bind('shift', () => this.setState({dir: -1}));
+    this.keys.bind('shift', () => this.setState({dir: 1}), 'keyup');
+
+    // Close the searchbox
+    this.keys.bind('esc', this.props.toggleSearch);
+    this.keys.bind('mod+f', this.props.toggleSearch);
+  }
+
+  componentWillUnmount() {
+    this.keys.unbind('shift');
+    this.keys.unbind('esc');
+    this.keys.unbind('mod+f');
+
+    const term = this.props.terms && this.props.terms.getActiveTerm();
+    if (term) {
+      term.focus();
+    }
+  }
+
+  template(css) {
+    return (<form
+      className={css('searchForm', this.state.searchError && 'noResults')}
+      onSubmit={this.handleSubmit}
+      ref={form => {
+        this.form = form;
+      }}
+      >
+      { this.props.customChildrenBefore }
+
+      <input
+        type="text"
+        role="search"
+        placeholder="Search..."
+        className={css('searchInput')}
+        style={{fontFamily: this.props.uiFontFamily}}
+        ref={input => {
+          if (input) {
+            this.searchBox = input;
+            this.searchBox.focus();
+          }
+        }}
+        value={this.state.searchTerm}
+        onChange={this.handleSearchTermChange}
+        />
+      { this.state.searchError ?
+        <div
+          className={css('noResultsText')}
+          style={{fontFamily: this.props.uiFontFamily}}
+          >
+          (no results)
+        </div> :
+        <input
+          type="submit"
+          value="Find"
+          className={css('searchButton')}
+          style={{fontFamily: this.props.uiFontFamily}}
+          />
+      }
+
+      { this.props.customChildren }
+    </form>);
+  }
+
+  styles() {
+    return {
+      searchForm: {
+        position: 'absolute',
+        top: '50px',
+        right: '50px',
+        width: '400px',
```

Comment:
Yeah, you're right. Do you know of any examples in the codebase where logic like this happens? If not off the top of your head, no worries - I'll dig around